shotover-proxy/src/transforms/protect/mod.rs

```diff
@@ -30,6 +33,161 @@ pub struct Protect {
+    /// encodes a Protected object into a byte array.  This is here to centeralize the serde for
+    /// the Protected object.
+    /// Returns an error if a Plaintext Protected object is passed
+    fn encode(protected: &Protected) -> Result<Vec<u8>> {
+        match protected {
+            Protected::Plaintext(_) => Err(anyhow!("can not encode plain text")),
+            Protected::Ciphertext { .. } => Ok(serde_json::to_vec(protected)?),
+        }
+    }
+
+    /// decodes a byte array into the Protected object.  This is here to centeralize the serde for
+    /// the Protected object.
+    fn decode(data: &[u8]) -> Result<Protected> {
+        Ok(serde_json::from_slice(data)?)
+    }
+
+    /// get the list of protected columns for the specified table name.  Will return `None` if no columns
+    /// are defined for the table.
+    fn get_protected_columns(&self, table_name: &FQName) -> Option<&Vec<String>> {
+        // TODO replace "" with cached keyspace name
+        if let Some(tables) = self
+            .keyspace_table_columns
+            .get(table_name.extract_keyspace(""))
+        {
+            tables.get(&table_name.name)
+        } else {
+            None
+        }
+    }
+
+    /// extractes the protected object from the message value.  Resulting object is a Protected::Ciphertext
+    fn extract_protected(&self, value: &MessageValue) -> Result<Protected> {
+        match value {
+            MessageValue::Bytes(b) => Protect::decode(&b[..]),
+            MessageValue::Varchar(hex_value) => {
+                let byte_value = hex::decode(hex_value);
+                Protect::decode(&byte_value.unwrap())
+            }
+            _ => Err(anyhow!(
+                "Could not get bytes to decrypt - wrong value type {:?}",
+                value
+            )),
+        }
+    }
+
+    /// determines if columns in the CassandraStatement need to be encrypted and encrypts them.  Returns `true` if any columns were changed.
+    ///  * `statement` the statement to encrypt.
+    ///  * `columns` the column names to encrypt.
+    ///  * `key_source` the key manager with encryption keys.
+    ///  * `key_id` the key within the manager to use.
+    async fn encrypt_columns(&self, statement: &mut CassandraStatement) -> Result<bool> {
+        let mut data_changed = false;
+        if let Some(table_name) = cql_statement::get_table_name(statement) {
+            if let Some(columns) = self.get_protected_columns(table_name) {
+                match statement {
+                    CassandraStatement::Insert(insert) => {
+                        // get the indices of the inserted protected columns
+                        let indices: Vec<usize> = insert
+                            .columns
+                            .iter()
+                            .enumerate()
+                            .filter_map(|(i, col_name)| {
+                                if columns.contains(col_name) {
+                                    Some(i)
```

Comment:
I wonder if we can just do the work immediately here without creating an intermediate vec